KAFKA-13773: catch kafkaStorageException to avoid broker shutdown directly

showuon opened the pull request:
[jira](https://issues.apache.org/jira/browse/KAFKA-13773) 

When logManager startup and loadLogs, we expect to catch any `IOException` (ex: out of space error) and turn the log dir into offline. Later, we'll handle the offline logDir in `ReplicaManager` [here](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/ReplicaManager.scala#L1914), so that the `cleanShutdown` file won't be created when all logDirs are offline. The reason why the broker shutdown with cleanShutdown file after full disk is because during loadLogs and do log recovery, we'll write `leader-epoch-checkpoint` file [here](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/log/LogLoader.scala#L184). And if any IOException thrown, we'll wrap it as `KafkaStorageException` and rethrow [here](https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/checkpoints/CheckpointFileWithFailureHandler.scala#L37-L42). And since we don't catch `KafkaStorageException`, so the exception is caught in the other place and go with clean shutdown path.

This PR is to fix the issue by catching the `KafkaStorageException` with `IOException` cause exceptions during loadLogs, and mark the logDir as offline to let the `ReplicaManager` handle the offline logDirs.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

showuon committed:
KAFKA-13773: catch kafkaStorageException to avoid broker shutdown unexpectedly

showuon commented:
@junrao , please take a look. Thanks.

junrao commented on core/src/test/scala/unit/kafka/log/LogLoaderTest.scala:
```diff
@@ -158,22 +197,54 @@ class LogLoaderTest {
+      logManager.shutdown()
+    }
+
+    locally {
```
Could we avoid duplicating the code by iterating the same logic twice?

showuon replied:
Updated. Thanks.

junrao commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -376,8 +381,10 @@ class LogManager(logDirs: Seq[File],
+                handleIOException(logDirAbsolutePath, e)
+              case e: KafkaStorageException if e.getCause.isInstanceOf[IOException] =>
+                // KafkaStorageException might be thrown, ex: during writing LeaderEpochFileCache
+                handleIOException(logDirAbsolutePath, e.getCause.asInstanceOf[IOException])
```
If we hit KafkaStorageException, that means the failed disk has already been reported to the logDirFailureChannel when KafkaStorageException was generated. So, we probably don't need to track it in offlineDirs again.

showuon replied:
Good catch! Updated.

junrao reviewed (commented):
@showuon : Thanks for the PR. A couple of comments.

showuon committed:
Merge branch 'trunk' of https://github.com/apache/kafka into KAFKA-13773

showuon committed:
KAFKA-13773: don't add offline logdir when kafkaStorageException

showuon commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -830,7 +830,13 @@ class KafkaServer(
   private def checkpointBrokerMetadata(brokerMetadata: ZkMetaProperties) = {
     for (logDir <- config.logDirs if logManager.isLogDirOnline(new File(logDir).getAbsolutePath)) {
       val checkpoint = brokerMetadataCheckpoints(logDir)
-      checkpoint.write(brokerMetadata.toProperties)
+      try {
+        checkpoint.write(brokerMetadata.toProperties)
+      } catch {
+      case e: IOException =>
+        val dirPath = checkpoint.file.getAbsolutePath
+        logDirFailureChannel.maybeAddOfflineLogDir(dirPath, s"Error while writing meta.properties to $dirPath", e)
```
handle IOException during writing meta.properties when broker startup, to avoid it gracefully shutdown the broker.

showuon commented:
@junrao , thanks for your review. I've addressed your comments. Also, I found we should handle `IOException` during writing `meta.properties` in server.startup. Thanks.

junrao commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -376,7 +381,9 @@ class LogManager(logDirs: Seq[File],
-                offlineDirs.add((logDirAbsolutePath, e))
+                handleIOException(logDirAbsolutePath, e)
+              case e: KafkaStorageException if e.getCause.isInstanceOf[IOException] =>
+                // KafkaStorageException might be thrown, ex: during writing LeaderEpochFileCache
```
When converting IOException to KafkaStorageException, we already add an error log. So, there is probably no need to log it below again. Could we also add a comment why we can just let go of KafkaStorageException?

showuon replied:
Yes, good suggestion.

junrao commented on core/src/test/scala/unit/kafka/log/LogLoaderTest.scala:
```diff
@@ -134,10 +154,29 @@ class LogLoaderTest {
+                                                 errorType: ErrorTypes.Errors = null,
+                                                 logDirFailureChannel: LogDirFailureChannel = new LogDirFailureChannel(logDirs.size)
+                                                ): (LogManager, Executable) = {
+      simulateError.hasError = hasError
```
We are changing simulateError after initializing LogManager. Perhaps it's clearer to just do that for the first test and avoid passing in hasError and errorType to initializeLogManagerForSimulatingErrorTest()?

junrao reviewed (commented):
@showuon : Thanks for the updated PR. A couple of more comments.

showuon committed:
KAFKA-13773: avoid output error message again

showuon commented:
@junrao , I've updated the PR in this commit: https://github.com/apache/kafka/pull/12136/commits/ca8cb7df7b7fbd17d95e7a34525840b5578fe520. Please take a look again. Thanks.

junrao commented on core/src/test/scala/unit/kafka/log/LogLoaderTest.scala:
```diff
@@ -158,22 +192,37 @@ class LogLoaderTest {
-        val defaultConfig = logManager.currentDefaultConfig
-        logManager.loadLogs(defaultConfig, logManager.fetchTopicConfigOverrides(defaultConfig, Set.empty))
-      })
+      // Simulate Runtime error
```
Should this comment be moved to before line 196?

junrao reviewed (commented):
@showuon : Thanks for the updated PR. Just a minor comment. Also, are the test failures related?

showuon committed:
Merge branch 'trunk' of https://github.com/apache/kafka into KAFKA-13773

alok123t commented on core/src/main/scala/kafka/log/LogManager.scala:
```diff
@@ -318,6 +318,11 @@ class LogManager(logDirs: Seq[File],
     val jobs = ArrayBuffer.empty[Seq[Future[_]]]
     var numTotalLogs = 0
 
+    def handleIOException(logDirAbsolutePath: String, e: IOException) = {
```
nit: return type unit

alok123t commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -830,7 +830,13 @@ class KafkaServer(
+      try {
+        checkpoint.write(brokerMetadata.toProperties)
+      } catch {
+      case e: IOException =>
```
nit: fix indentation 
missing tab before case statement

alok123t reviewed (commented):
few minor nits

showuon committed:
KAFKA-13773: remove testCleanShutdownAfterFailedStartupDueToCorruptLogs test due to wrong assumption

showuon committed:
Merge branch 'trunk' of https://github.com/apache/kafka into KAFKA-13773

showuon committed:
KAFKA-13773: update the test to make sure it exit as expected

showuon commented:
@junrao , thanks for the comment. I've updated the PR. The failed test: `testCleanShutdownAfterFailedStartupDueToCorruptLogs` is indeed related. But the test is not the correct expectation. When log corrupted, we should not clean shutdown and expect to recover logs in next start up. I updated the test.

@alok123t , thanks for the comment. I've updated the PR. Thanks.

junrao commented on core/src/test/scala/unit/kafka/server/ServerShutdownTest.scala:
```diff
@@ -158,14 +158,29 @@ class ServerShutdownTest extends KafkaServerTestHarness {
-    verifyCleanShutdownAfterFailedStartup[KafkaStorageException]
+
+    val expectedStatusCode = Some(1)
+    var receivedStatusCode = Option.empty[Int]
```
Should receivedStatusCode be volatile since it's set in a different thread?

showuon replied:
Good call! Thanks.

junrao commented on core/src/test/scala/unit/kafka/server/ServerShutdownTest.scala:
```diff
@@ -158,14 +158,29 @@ class ServerShutdownTest extends KafkaServerTestHarness {
+      // this startup should fail with no online log dir (due to corrupted log), and exit directly without throwing exception
+      assertDoesNotThrow(recreateBrokerExec)
+      // JVM should exit with status code 1
+      assertEquals(expectedStatusCode, receivedStatusCode)
```
LogDirFailureHandler sets the exit flag but runs in a separate thread. Should we do this in a waitUntil() block?

showuon replied:
Good call again! Thanks.

junrao commented on core/src/test/scala/unit/kafka/server/ServerShutdownTest.scala:
```diff
@@ -158,14 +158,29 @@ class ServerShutdownTest extends KafkaServerTestHarness {
+      // JVM should exit with status code 1
+      assertEquals(expectedStatusCode, receivedStatusCode)
+    } finally {
+      Exit.resetExitProcedure()
```
Should this be resetHaltProcedure?

showuon replied:
Oops! Thanks.

junrao reviewed (commented):
@showuon : Thanks for finding the test issue. A few more comments below.

showuon committed:
KAFKA-13773: fix test issue

junrao reviewed (approved):
@showuon : Thanks for the updated PR. LGTM

showuon commented:
@junrao ,Thank you for your patient review!

Failed tests are unrelated.
```
Build / JDK 11 and Scala 2.13 / kafka.controller.ControllerIntegrationTest.testTopicIdUpgradeAfterReassigningPartitions()
```

showuon commented:
cherry-pick back to 3.2 branch
